docs/rfc/0001-exporter-interface.md

```diff
@@ -0,0 +1,194 @@
+interface and extend functionality to new data outputs, we will be unable to implement all of them and maintain them
+adequately. And any complex data outputs such as those mentioned above will be blocked--users cannot easily use the
+existing interface which Indexer exposes internally to customize their data output.
+
+Once we have a fully implemented exporter plugin interface, users of Indexer will be able to easily develop and
+run plugins which encapsulate their own business logic for data destinations. Some of these will be generic enough that
+we will maintain them centrally in our Indexer repository. Others will be specific to a user's needs and will be
+maintained either publicly or privately  by them. This plugin specification will also give us a runtime framework and
+configuration necessary for creating other plugin systems, such as input plugins and intermediate plugins.
+
+## Design proposal
+
+[design-proposal]: #design-proposal
+
+### Indexer-Native Exporter Plugins  
+Exporter plugins that are native to the Indexer (maintained within the Indexer repository) will necessarily behave
+differently than those which are maintained and built separately and loaded into the pipeline at runtime. The following
+section will outline the properties of so called "Indexer-native" plugins, which will then be extended in the subsequent
+section to incorporate non-native plugins.
+
+***Plugin Interface***  
+Exporter plugins that are native to the Indexer (maintained within the Indexer repository) will each have the
+following components:
+* An implementation of the exporter interface.
+```
+// Exporter defines the methods invoked during the plugin lifecycle of a data exporter.
+type Exporter interface {
+  // Connect will be called during initialization, before block data starts going through the pipeline.
+  // Typically used for things like initializating network connections.
+  // The ExporterConfig passed to Connect will contain the Unmarhsalled config file specific to this plugin.
+  // Should return an error if it fails--this will result in the Indexer process terminating.
+  Connect(ExporterConfig cfg) error
+  
+  // Shutdown will be called during termination of the Indexer process.
+  // There is no guarantee that plugin lifecycle hooks will be invoked in any specific order in relation to one another.
+  // Returns an error if it fails which will be surfaced in the logs, but the process is already terminating.
+  Shutdown() error
+  
+  // Recv is called for each block to be processed by the exporter.
+  // Should return an error on failure--retries are configurable.
+  Recv(DataFormat blockData) error
+  
+  // DataFormat returns the type of the data that the plugin expects to receive.
+  // Must be one of the impelemented formats in the Indexer repository.
+  DataFormat() DataFormat
+  
+  // Round returns the next round to be processed. Atomically updated when Recv successfully completes.
+  Round() uint64
```

Comment:
This is worded a little strongly. It suggests that the exporter could choose to re-process old rounds. That may be a feature we want to support eventually, but I'm not sure it's part of v1.
```suggestion
  // Round returns the next round that this exporter has not yet successfully processed.
  Round() uint64
```